### DISCUSSION TITLE
Parse nested Json, normalize it and save as parquet

### DISCUSSION THREAD

**Discussion by aser86** [duckdb/duckdb#10443] [Q&A]
I need help, because i despair on this json format and can not parse it with duckdb.
I trieyd to read this json with read_ndjson_auto function but always get the struct of the files.
How I can get the key and value entries?
How I can normalize this format in the under table structure?

My duckdb version: v0.9.2

JSON-File
`
{
	"system": 123,
	"time" : "2023-10-01",
	"files": {
		"fileA": {
			"name": "myfile",
			"size": 118,
			"value": "text"
		},
		"fileB": {
			"name": "myfile2",
			"size": 244,
			"value": "my text"
		},
		"fileC": {
			"name": "myfile3",
			"size": 71,
			"value": ""
		}
	}
}
`

My Goal is atable which i want to save in parquet.

system              time             files            name            size            value
123                   2023-10-01  fileA           myfile           118              text
123                   2023-10-01  fileB           myfile2          244             my text
123                   2023-10-01  fileC           myfile3          71

**Comment by cmdlineluser** [comment]
**Disclaimer:** I'm very much a DuckDB novice, so there may be a better way to do this.

If the `system`, `time` and `files` keys are known, you could:

- unpack files using the  [`files.*`](https://duckdb.org/docs/sql/data_types/struct.html#struct) struct syntax
- [UNPIVOT](https://duckdb.org/docs/sql/statements/unpivot) into rows

```python
duckdb.sql("""
unpivot 
   (from 'filename.json' select system, time, files.*)
on 
   * exclude(system, time)
into 
   name files 
   value metadata
""")
```

```python
# ┌────────┬────────────┬─────────┬───────────────────────────────────────────────┐
# │ system │    time    │  files  │                   metadata                    │
# │ int64  │    date    │ varchar │ struct("name" varchar, size bigint, "value"…  │
# ├────────┼────────────┼─────────┼───────────────────────────────────────────────┤
# │    123 │ 2023-10-01 │ fileA   │ {'name': myfile, 'size': 118, 'value': text}  │
# │    123 │ 2023-10-01 │ fileB   │ {'name': myfile2, 'size': 244, 'value': my …  │
# │    123 │ 2023-10-01 │ fileC   │ {'name': myfile3, 'size': 71, 'value': }      │
# └────────┴────────────┴─────────┴───────────────────────────────────────────────┘
```

You can then:

- [UNNEST](https://duckdb.org/docs/sql/query_syntax/unnest) the "metadata"

```python
duckdb.sql("""
from (
   unpivot 
      (from 'filename.json' select system, time, files.*)
   on 
      * exclude(system, time)
   into 
      name files 
      value metadata
)
select * exclude metadata, unnest(metadata)
""")
```

```python
# ┌────────┬────────────┬─────────┬─────────┬───────┬─────────┐
# │ system │    time    │  files  │  name   │ size  │  value  │
# │ int64  │    date    │ varchar │ varchar │ int64 │ varchar │
# ├────────┼────────────┼─────────┼─────────┼───────┼─────────┤
# │    123 │ 2023-10-01 │ fileA   │ myfile  │   118 │ text    │
# │    123 │ 2023-10-01 │ fileB   │ myfile2 │   244 │ my text │
# │    123 │ 2023-10-01 │ fileC   │ myfile3 │    71 │         │
# └────────┴────────────┴─────────┴─────────┴───────┴─────────┘
```

**Comment by aser86** [comment]
Thank you this helps for the example file but my real file has about 2.000 entries. With this file i dont come to an end with this sql.

I made an test file like this with only 500 entries and the issue is thas the query take a lot of memory  and runs about 5 minutes.
Is this a known issue?

The code:
> unpivot 
   (from '0_test_memory.json' select system, time, files.*)
on  * exclude(system, time)
into 
   name files 
   value metadata;

Here my file example:
[Testfile](https://github.com/aser86/test/blob/main/0_test_memory.json)

**Comment by cmdlineluser** [comment]
> Is this a known issue?

I'm not sure.

It runs for me in 15/20 seconds - but it does use a lot of RAM. (~10GB)

```shell
$ python3 -m pip install --pre duckdb
# Successfully installed duckdb-0.9.3.dev3715
```

```python
>>> %%time
... duckdb.sql("""
... from (
...    unpivot
...       (from '0_test_memory.json' select system, time, files.*)
...    on
...       * exclude(system, time)
...    into
...       name files
...       value metadata
... )
... select * exclude metadata, unnest(metadata)
... """).pl()

# shape: (500, 6)
# ┌────────┬────────────┬──────────┬───────────┬───────┬─────────┐
# │ system ┆ time       ┆ files    ┆ name      ┆ size  ┆ value   │
# │ ---    ┆ ---        ┆ ---      ┆ ---       ┆ ---   ┆ ---     │
# │ i64    ┆ date       ┆ str      ┆ str       ┆ i64   ┆ str     │
# ╞════════╪════════════╪══════════╪═══════════╪═══════╪═════════╡
# │ 123    ┆ 2023-10-01 ┆ fileA1   ┆ myfile1   ┆ 100   ┆ text1   │
# │ 123    ┆ 2023-10-01 ┆ fileA2   ┆ myfile2   ┆ 200   ┆ text2   │
# │ 123    ┆ 2023-10-01 ┆ fileA3   ┆ myfile3   ┆ 300   ┆ text3   │
# │ 123    ┆ 2023-10-01 ┆ fileA4   ┆ myfile4   ┆ 400   ┆ text4   │
# │ 123    ┆ 2023-10-01 ┆ fileA5   ┆ myfile5   ┆ 500   ┆ text5   │
# │ …      ┆ …          ┆ …        ┆ …         ┆ …     ┆ …       │
# │ 123    ┆ 2023-10-01 ┆ fileA496 ┆ myfile496 ┆ 49600 ┆ text496 │
# │ 123    ┆ 2023-10-01 ┆ fileA497 ┆ myfile497 ┆ 49700 ┆ text497 │
# │ 123    ┆ 2023-10-01 ┆ fileA498 ┆ myfile498 ┆ 49800 ┆ text498 │
# │ 123    ┆ 2023-10-01 ┆ fileA499 ┆ myfile499 ┆ 49900 ┆ text499 │
# │ 123    ┆ 2023-10-01 ┆ fileA500 ┆ myfile500 ┆ 50000 ┆ text500 │
# └────────┴────────────┴──────────┴───────────┴───────┴─────────┘
```

```
Elapsed time: 20.86230 seconds
Memory usage: 9.8 GB
```

**Comment by cmdlineluser** [comment]
Even simplifying the query to a single `UNNEST` has similar performance:

```python
%%time
x = duckdb.sql("from '0_test_memory.json' select unnest(files)").pl()

# Elapsed time: 19.04648 seconds
# Memory usage: 9.7 GB
```

For comparison, doing the operation directly in Polars runs instantly:

```python
%time
y = pl.read_json("0_test_memory.json").unnest("files")

# Elapsed time: 0.01538 seconds
# Memory usage: 7.9 MB
```

I think this single UNNEST example probably needs to be reported as a performance issue?

**[EDIT]:** Appears to be a known issue.

- https://github.com/duckdb/duckdb/issues/9628
